src/traces/ohlc/calc.js

```diff
@@ -0,0 +1,163 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Lib = require('../../lib');
+var _ = Lib._;
+var Axes = require('../../plots/cartesian/axes');
+
+function calc(gd, trace) {
+    var xa = Axes.getFromId(gd, trace.xaxis);
+    var ya = Axes.getFromId(gd, trace.yaxis);
+
+    var tickLen = convertTickWidth(gd, xa, trace);
+    var minDiff = trace._minDiff;
+    delete trace._minDiff;
+    var x = trace._xcalc;
+    delete trace._xcalc;
+
+    var cd = calcCommon(gd, trace, x, ya, ptFunc);
+
+    Axes.expand(xa, x, {vpad: minDiff / 2});
+
+    if(cd.length) {
+        Lib.extendFlat(cd[0].t, {
+            wHover: minDiff / 2,
+            tickLen: tickLen
+        });
+        return cd;
+    } else {
+        return [{t: {empty: true}}];
+    }
+}
+
+function ptFunc(o, h, l, c) {
+    return {
+        o: o,
+        h: h,
+        l: l,
+        c: c
+    };
+}
+
+
+// shared between OHLC and candlestick
+// ptFunc makes a calcdata point specific to each trace type, from oi, hi, li, ci
+function calcCommon(gd, trace, x, ya, ptFunc) {
+    var o = ya.makeCalcdata(trace, 'open');
+    var h = ya.makeCalcdata(trace, 'high');
+    var l = ya.makeCalcdata(trace, 'low');
+    var c = ya.makeCalcdata(trace, 'close');
+
+    var hasTextArray = Array.isArray(trace.text);
+
+    // we're optimists - before we have any changing data, assume increasing
+    var increasing = true;
+    var cPrev = null;
+
+    var cd = [];
+    for(var i = 0; i < x.length; i++) {
+        var xi = x[i];
+        var oi = o[i];
+        var hi = h[i];
+        var li = l[i];
+        var ci = c[i];
+
+        if(xi !== undefined && oi !== undefined && hi !== undefined && li !== undefined && ci !== undefined) {
```

Comment:
Hmm. you mean `BADNUM` here?

Reply:
good call, yes `BADNUM` -> https://github.com/plotly/plotly.js/pull/2561/commits/dc016858219587070e633f32b248ae5497e384b4